src/internal/observable/from.ts

```diff
@@ -119,34 +119,130 @@ export function from<O extends ObservableInput<any>>(input: O, scheduler: Schedu
-  if (!scheduler) {
-    if (input instanceof Observable) {
-      return input;
+  return scheduler ? scheduled(input, scheduler) : innerFrom(input);
+}
+
+// TODO: Use this throughout the library, rather than the `from` above, to avoid
+// the unnecessary scheduling check and reduce bundled sizes of operators that use `from`.
+// TODO: Eventually, this just becomes `from`, as we don't have the deprecated scheduled path anymore.
+export function innerFrom<T>(input: ObservableInput<T>) {
+  if (input instanceof Observable) {
+    return input;
+  }
+  if (input != null) {
+    if (isInteropObservable(input)) {
+      return fromInteropObservable(input);
+    }
+    if (isArrayLike(input)) {
+      return fromArrayLike(input);
+    }
+    if (isPromise(input)) {
+      return fromPromise(input);
+    }
+    if (isIterable(input)) {
+      return fromIterable(input as any);
+    }
+    if (isAsyncIterable(input)) {
+      return fromAsyncIterable(input as any);
     }
-    return new Observable<T>(subscribeTo(input));
-  } else {
-    return scheduled(input, scheduler);
   }
+
+  throw createInvalidObservableTypeError(input);
+}
+
+/**
+ * Creates an RxJS Observable from an object that implements `Symbol.observable`.
+ * @param obj An object that properly implements `Symbol.observable`.
+ */
+function fromInteropObservable<T>(obj: any) {
+  return new Observable((subscriber: Subscriber<T>) => {
+    const obs = (obj as any)[Symbol_observable]();
+    if (typeof obs.subscribe !== 'function') {
+      // Should be caught by observable subscribe function error handling.
+      throw new TypeError('Provided object does not correctly implement Symbol.observable');
+    } else {
+      return obs.subscribe(subscriber);
+    }
+  });
+}
+
+/**
+ * Synchronously emits the values of an array like and completes.
+ * This is exported because there are creation functions and operators that need to
+ * make direct use of the same logic, and there's no reason to make them run through
+ * `from` conditionals because we *know* they're dealing with an array.
+ * @param array The array to emit values from
+ */
+export function fromArrayLike<T>(array: ArrayLike<T>) {
+  return new Observable((subscriber: Subscriber<T>) => {
+    // Loop over the array and emit each value. Note two things here:
+    // 1. We're making sure that the subscriber is not closed on each loop.
+    //    This is so we don't continue looping over a very large array after
+    //    something like a `take`, `takeWhile`, or other synchronous unsubscription
+    //    has already unsubscribed.
+    // 2. In this form, reentrant code can alter that array we're looping over.
+    //    This is a known issue, but considered an edge case. The alternative would
+    //    be to copy the array before executing the loop, but this has
+    //    performance implications.
+    for (let i = 0; i < array.length && !subscriber.closed; i++) {
```

Comment:
question: Would there be any advantage in using `for ... of` here? 

```ts
for (const value of array) {
  if (subscriber.closed) break;
  subscriber.next(value);
}
```

Reply:
The primary reason is that this works for any ArrayLike. `for...of` will not.

Also, `for...of`, last I checked, was notably slower than the `for(;;)` loop, as it uses the iterator instead of accessing the array by index. However, there's an argument to be made for condensing the `fromIterable` and `fromArray` code into one thing, but again, that doesn't work for every ArrayLike. We'd have to do an `Array.from()` first, and then iterate it, which would cause double iteration and slow us down.